orttraining/orttraining/training_ops/cuda/communication/nccl_service.h

```diff
@@ -0,0 +1,185 @@
+
+namespace onnxruntime {
+namespace cuda {
+
+struct NcclTask final {
+  // Attributes for communication operator.
+  enum class Type { SEND,
+                    RECV,
+                    ALLREDUCE };
+  // Operator to perform.
+  Type type;
+  // For Send, this field is destination device's ID.
+  // For Recv, this field is source device's ID.
+  std::vector<int> peers;
+
+  // Attributes for memory location.
+  // GPU memory pointer.
+  void* ptr;
+  // Number of bytes to send/recv.
+  size_t size;
+
+  // Scheduler flag.
+  bool is_enqueued;
+  bool is_finished;
+
+  // Debug information
+  std::string info;
+
+  // Return true if the two operations are the same.
+  bool Compare(const NcclTask& other) const;
+
+  // Clear runtime information.
+  void ResetTask();
+};
+
+// A collection of independent communication operations.
+struct NcclTaskGroup final {
+  // Schedule a communication operation in this group.
+  // We don't know the pointer to the actual data and other runtime information yet;
+  // runtime information is filled by calling EqunueTask(...).
+  void PlanTask(const NcclTask::Type type, const std::vector<int> peers);
+  // Fill in task's details.
+  const NcclTask* EqueueTask(
+      const NcclTask::Type type,
+      const std::vector<int> peers,
+      void* ptr,
+      const size_t size,
+      const std::string info);
+  bool IsAllTasksEqueued() const;
+  bool IsAllTasksFinished() const;
+  void ResetAllTasks();
+  friend std::ostream& operator<<(std::ostream& stream, const NcclTaskGroup& task_group);
+  std::vector<NcclTask> batch;
+};
+
+// The use of this class has two stages. First, the user needs to plan the communication operators.
+// Second, when running a model, the user should submit tasks following the communication plan.
+// Function names begin with "Plan" are used for creating communication plan. Function names begin
+// with "Submit" asks this class to run the submitted task. Communication usually does not happen
+// immediately after submitting a task. The actual communication time is decided by this class based on
+// the communication plan.
+//
+// Below is an example of planning tasks. Notice that the communication operations in the same group are
+// called in random order, so those operations cannot have mutual dependency.
+//
+//   auto& nccl_service = cuda::NcclService::GetInstance();
+//
+//   nccl_service.PlanStart();         // Signal the begin of communication planning.
+//
+//   nccl_service.PlanStartNewGroup(); // Create new time slot.
+//   nccl_service.PlanSend(0);
+//   nccl_service.PlanRecv(1);
+//   nccl_service.PlanEndNewGroup();   // Mark the end of the first time slot.
+//
+//   nccl_service.PlanStartNewGroup(); // Create the second time slot.
+//   nccl_service.PlanSend(1);
+//   nccl_service.PlanRecv(0);
+//   nccl_service.PlanEndNewGroup();   // Mark the end of the second time slot.
+//
+//   nccl_service.EndPlan();           // Signal the end of communication planning.
```

Comment:
>nccl_service.EndPlan();           // Signal the end of communication planning. [](start = 5, length = 78)

Maybe rename this to PlanEnd() for consistency -- "Function names begin with "Plan" are used for creating communication plan."